module/localisation/RobotParticleLocalisation/src/RobotParticleLocalisation.cpp

```diff
@@ -64,48 +73,140 @@ namespace module::localisation {
-                                                   poss,
-                                                   goals.Hcw,
-                                                   m.type,
-                                                   fd);
-                                }
-                                else {
-                                    log("Received non-finite measurements from vision. Discarding ...");
+                    // Currently this has no idea about pairs of posts, only that it has been labelled left, right and
+                    // unknown
+
+                    // Save the filter state
+                    auto saved_filter = filter;
+
+                    // Go through all of the known sided posts
+                    for (auto goal_post : goals.goals) {
+                        if (goal_post.side != Goal::Side::UNKNOWN_SIDE) {
+                            // These are either left or right goal posts
+
+                            // Compare each of these to the possible goal posts on the field (own and opp)
+
+                            // TODO remove the repeated measurements
+                            auto m = goal_post.measurements[0];
+
+                            // Check that the measurement is sane
+                            if (m.position.allFinite() && m.covariance.allFinite()) {
+                                auto filter_new_own = saved_filter;
+
+                                // Make a new candidate for the post
+                                auto candidate_own =
+                                    filter_new_own.measure(Eigen::Vector3d(m.position.cast<double>()),
+                                                           Eigen::Matrix3d(m.covariance.cast<double>()),
+                                                           getFieldPosition(goal_post, fd, 1),  // Own post
+                                                           goals.Hcw);                          //,
+                                                                                                //    m.type,
+                                                                                                //    fd);
+
+                                auto filter_new_opp = saved_filter;
+
+                                // Make a new candidate for the post
+                                auto candidate_opp =
+                                    filter_new_opp.measure(Eigen::Vector3d(m.position.cast<double>()),
+                                                           Eigen::Matrix3d(m.covariance.cast<double>()),
+                                                           getFieldPosition(goal_post, fd, 0),  // Opp post
+                                                           goals.Hcw);                          //,
+                                //    m.type,
+                                //    fd);
+                                if (config.debug) {
```

Comment:
That is unfortunate, what would you prefer?

Reply:
But why are you not happy?
Either way I would suggest that rather than having a debug true/false you should just take advantage of the log level we already have
```suggestion
                                if (log_level <= NUClear::DEBUG) {
```